fs/test.py

```diff
@@ -1494,6 +1494,10 @@ def test_upload(self):
         with self.fs.open("foo", "rb") as f:
             data = f.read()
         self.assertEqual(data, b"bar")
+        
+        # upload to non-existing path (/foo/bar)
+        with self.assertRaises(errors.ResourceNotFound):
+            self.fs.upload("/foo/bar/baz", bytes_file)
```

Comment:
Ok, I'm clearly nitpicking here, but in that test isn't `/foo` an existing file here? There is a risk that a different error is raised by some implementors (`DirectoryExpected` for instance).

Could you try with a path that truly doesn't exist (e.g. `/spam/eggs`) ?

Reply:
I somehow missed that ;-) Maybe it is even more sufficient to assert calling underlying `openbin`, since that already has tests that assert raising the `ResourceNotFound`

Reply:
Actually, some filesystem may have a more efficient method that does *not* use `openbin`, so only checking `openbin` would not be a complete test.

E.g. for my [`fs.sshfs`](https://github.com/althonos/fs.sshfs) implementation, I ~~could~~ **should** implement the `SSHFS.upload` method in a way that uses [`SFTP.putfo`](http://docs.paramiko.org/en/stable/api/sftp.html#paramiko.sftp_client.SFTPClient.putfo) to upload a file rather than using the default method `openbin`, which calls [`SFTP.open`](http://docs.paramiko.org/en/stable/api/sftp.html#paramiko.sftp_client.SFTPClient.open)!